cpp/segtree.hpp

```diff
@@ -0,0 +1,336 @@
+    /**
+     * @brief 指定された要素の値にxを作用させる
+     * 
+     * @param k インデックス
+     * @param x 作用素
+     */
+    void apply(int k, S& x) { set(k, ActualSegTree::op(get(k), x)); }
+    
+    /**
+     * @brief [l, r)の区間の総積を返す
+     * 
+     * @param l 半開区間の開始
+     * @param r 半開区間の終端
+     * @return S 総積
+     */
+    S prod(int l, int r) const {
+        S left_prod = ActualSegTree::e(), right_prod = ActualSegTree::e();
+        l += sz;
+        r += sz;
+        while (l < r) {
+            if (l & 1) left_prod = ActualSegTree::op(left_prod, data[l++]);
+            if (r & 1) right_prod = ActualSegTree::op(data[--r], right_prod);
+            l >>= 1;
+            r >>= 1;
+        }
+        return ActualSegTree::op(left_prod, right_prod);
+    }
+    /**
+     * @brief すべての要素の総積を返す
+     * 
+     * @return S 総積
+     */
+    S all_prod() const { return data[1]; }
+
+    /**
+     * @brief f(prod([l, r))) = trueとなる最大のrを返す
+     * 
+     * @tparam F
+     * @param l 半開区間の開始
+     * @param f 判定関数
+     * @return int 最大のr
+     */
+    template <typename F>
+    int max_right(int l, F f) const {
+        assert(f(ActualSegTree::e()));
+        if (l == n) return n;
+        l += sz;
+        while (l % 2 == 0) l >>= 1;
+        S sum = ActualSegTree::e();
+        while(f(ActualSegTree::op(sum, data[l]))) {
+            if (__builtin_clz(l) != __builtin_clz(l+1)) return n;
+            sum = ActualSegTree::op(sum, data[l]);
+            l++;
+            while (l % 2 == 0) l >>= 1;
+        }
+        while (l < sz) {
+            if (!f(ActualSegTree::op(sum, data[l * 2]))) l *= 2;
+            else {
+                sum = ActualSegTree::op(sum, data[l * 2]);
+                l = l * 2 + 1;
+            }
+        }
+        return l - sz;
+    }
+    /**
+     * @brief f(prod([l, r))) = trueとなる最小のlを返す
+     * 
+     * @tparam F
+     * @param r 半開区間の終端
+     * @param f 判定関数
+     * @return int 最小のl
+     */
+    template <typename F>
+    int min_left(int r, F f) const {
+        assert(f(ActualSegTree::e()));
+        if (r == 0) return 0;
+        r += sz - 1;
+        while (r % 2 == 1) r >>= 1;
+        S sum = ActualSegTree::e();
+        while(f(ActualSegTree::op(sum, data[r]))) {
+            if (__builtin_clz(r) != __builtin_clz(r-1)) return 0;
+            sum = ActualSegTree::op(sum, data[r]);
+            r--;
+            while (r % 2 == 1) r >>= 1;
+        }
+        while (r < sz) {
+            if (!f(ActualSegTree::op(data[r * 2 + 1], sum))) r = r * 2 + 1;
+            else {
+                sum = ActualSegTree::op(data[r * 2 + 1], sum);
+                r *= 2;
+            }
+        }
+        return r + 1 - sz;
+    }
+};
+
+/**
+ * @brief 積のファンクタが静的な場合のセグメント木の実装
+ * 
+ * @tparam S モノイドの型
+ * @tparam Op 積のファクタ
+ * @tparam E 積の単位元
+ */
+template <typename S, typename Op, S E>
+class StaticSegTree : public SegTreeBase<S, StaticSegTree<S, Op, E>> {
+    static Op operator_object;
```

Comment:
この変数の実体が存在せず、リンクエラーが起こり得ます。これは静的メンバ変数の**宣言**で、別に**定義**が必要です。
C++17 では `inline` をつけるだけでこの問題を回避できます。

```suggestion
    inline static Op operator_object;
```

Reply:
インライン変数初耳でした:eyes:
理解しましたありがとうございます